thoth/storages/graph/postgres.py

```diff
@@ -4922,6 +4922,59 @@ def sync_dependency_monkey_result(self, document: dict) -> None:
 
                 inspection_run.dependency_monkey_run_id = dependency_monkey_run.id
 
+    def get_python_package_required_symbols(
+        self, package_name: str, package_version: str, index_url: str
+    ) -> Optional[List[str]]:
+        """Get required symbols for a Python package in a specified version."""
+        package_name = self.normalize_python_package_name(package_name)
+        package_version = self.normalize_python_package_version(package_version)
+
+        with self._session_scope() as session:
+            query = session.query(PythonPackageVersion)
+
+            query = (
+                query.filter(PythonPackageVersion.package_name == package_name)
+                .filter(PythonPackageVersion.package_version == package_version)
+                .join((HasArtifact, PythonPackageVersionEntity.python_artifacts))
+                .join((PythonArtifact, HasArtifact.python_artifact))
+                .with_entities(PythonArtifact.versioned_symbols)
+            )
+
+            return query.first()   # If query result is empty returns None
+
+    def get_image_symbols(
+        self,
+        os_name: str,
+        os_version: str,
+        python_version: str,
+        *,
+        cuda_version: str,
+    ) -> Optional[List[str]]:
+        """Get symbols associated with a given image."""
+        with self._session_scope() as session:
+            if cuda_version:
+                query = (
+                    session.query(PackageExtractRun)
+                    .filter(PackageExtractRun.os_name == os_name)
+                    .filter(PackageExtractRun.os_version == os_version)
+                    .filter(PackageExtractRun.python_version == python_version)
+                    .filter(PackageExtractRun.cuda_version == cuda_version)
+                    .with_entities(
+                        PackageExtractRun.versioned_symbols
+                    ))
+            else:
+                query = (
+                    session.query(PackageExtractRun)
+                    .filter(PackageExtractRun.os_name == os_name)
+                    .filter(PackageExtractRun.os_version == os_version)
+                    .filter(PackageExtractRun.python_version == python_version)
+                    .filter(PackageExtractRun.cuda_version == cuda_version)
+                    .with_entities(
+                        PackageExtractRun.versioned_symbols
+                    ))
```

Comment:
IMHO we could reduce the code splitting the pieces of the queries and adding optional only one filter :) WDYT?

```
query = (
    session.query(PackageExtractRun)
    .filter(PackageExtractRun.os_name == os_name)
    .filter(PackageExtractRun.os_version == os_version)
    .filter(PackageExtractRun.python_version == python_version)
)  

if cuda_version:
    query = query.filter(PackageExtractRun.cuda_version == cuda_version)

query = query.with_entities(
    PackageExtractRun.versioned_symbols
    )
```

Reply:
I like this, makes functionality more apparent and cuts down on the function size.

Reply:
BTW this will change the query semantics.